### PR TITLE
Thread-safe refinement and stateful material property projection

### LINKED ISSUE #20154: materials/stateful_prop.adaptivity/spatially_varying occasionally crashes with threads
## Bug Description
Crashes occasionally crop up in CIVET, for example [here](https://civet.inl.gov/job/943137/)

## Steps to Reproduce
Run CIVET a bunch of times with threading. Additionally running with `ThreadSanitizer` there are indeed bugs related to `ProjectMaterialProperties`

## Impact
Impedes CI and could be causing users occasional random undefined behavior (like crashes).

### PR DESCRIPTION
**Pull request by lindsayad**
Closes #20154

### CODE REVIEW

**Comment by lindsayad** [comment]
@friedmud what do you think ... go with the extra element loop I added here or add locks to `ProjectMaterialProperties`?

**Comment by lindsayad** [comment]
I think the loop fission is fine so I think this is ready for review/merge

File: framework/src/problems/FEProblemBase.C [hunk 1]
```diff
@@ -6324,7 +6324,18 @@ FEProblemBase::meshChangedHelper(bool intermediate_change)
...
-      Threads::parallel_reduce(*_mesh.refinedElementRange(), pmp);
+      const auto & range = *_mesh.refinedElementRange();
+      Threads::parallel_reduce(range, pmp);
+
+      // Concurrent erasure from the shared hash map is not safe while we are reading from it in
+      // ProjectMaterialProperties, so we handle erasure here. Moreover, erasure based on key is not
+      // thread safe in and of itself because it is a read-write operation
+      for (auto it = range.begin(); it != range.end(); ++it)
+      {
+        const Elem * const elem = *it;
```
Comment (loganharbour) [inline framework/src/problems/FEProblemBase.C:6335]:
```suggestion
      for (const auto & elem : make_range(range.begin(), range.end())
      {
```

for future reference

**Review by loganharbour — DISMISSED** [review]

File: framework/src/problems/FEProblemBase.C (hunk 1 above)
Comment (dschwen) [reply framework/src/problems/FEProblemBase.C:6335]: ?! why make a range when you already have a range? couldn't you just do

```
      for (const auto & elem : range)
```

Comment (loganharbour) [reply framework/src/problems/FEProblemBase.C:6335]: oops

**Review by loganharbour — APPROVED** [review]

**Pull request merged**